### PR TITLE
add CapsNet example

### PR DESCRIPTION
**Pull request by Soonhwan-Kwon**
## Description ##
This example is MXNet implementation of [CapsNet](https://arxiv.org/abs/1710.09829):  
Sara Sabour, Nicholas Frosst, Geoffrey E Hinton. Dynamic Routing Between Capsules. NIPS 2017

We achieved `the best test error rate=0.29%` and `average test error=0.303%`. It is the best accuracy and fastest training time result among other implementations(Keras, Tensorflow at 2017-11-23).
The result on paper is `0.25% (average test error rate)`.

| Implementation| test err(%) | ※train time/epoch | GPU  Used|
| :---: | :---: | :---: |:---: |
| MXNet | 0.29 | 36 sec | 2 GTX 1080 |
| tensorflow | 0.49 | ※ 10 min | Unknown(4GB Memory) |
| Keras | 0.30 | 55 sec | 2 GTX 1080 Ti |

※ tensorflow implementation's batch size is 128 but MXNet and Keras implementation 's batchsize are 100. 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##

### CODE REVIEW

**Comment by Godricly** [comment]
Do you actually update your bias in capsule layer in each training iteration?

**Comment by Soonhwan-Kwon** [comment]
routingalgorithm

As stated in the routing algorithm of CapsNet (line 2),   
bias doesn't updated during routing and initialized to 0 for each batch(or training iteration),   
and it is implemented in that way.

File: example/capsnet/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+```
+On Multi gpus
+```
+python capsulenet.py --devices gpu0,gpu1
+```
+
+* * *
+## **Prerequisities**
+
+MXNet version above (0.11.0)
```
Comment (sxjscience) [inline example/capsnet/README.md]: For the version here, have you tested on some more recent versions like the current master or 1.0.0rc?

File: example/capsnet/capsulenet.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    x_recon = mx.sym.Activation(data=x_recon, act_type='relu', name='x_recon_act')
+    x_recon = mx.sym.FullyConnected(data=x_recon, num_hidden=1024, name='x_recon2')
+    x_recon = mx.sym.Activation(data=x_recon, act_type='relu', name='x_recon_act2')
+    x_recon = mx.sym.FullyConnected(data=x_recon, num_hidden=np.prod(input_shape), name='x_recon3')
+    x_recon = mx.sym.Activation(data=x_recon, act_type='sigmoid', name='x_recon_act3')
+
+    data_flatten = mx.sym.flatten(data=data)
+    squared_error = mx.sym.square(x_recon-data_flatten)
+    recon_error = mx.sym.mean(squared_error)
+    loss = mx.symbol.MakeLoss((1-0.392)*margin_loss(y_onehot, out_caps)+0.392*recon_error)
```
Comment (sxjscience) [inline example/capsnet/capsulenet.py]: I think it's better to make 0.392 as part of the function arguments.

**Comment by sxjscience** [comment]
@piiswrong Do you think this example can be merged in? It looks good.

**Comment by Soonhwan-Kwon** [comment]
@sxjscience Thank you for review, 
First we tested on MXNet 0.12.1 and it works well,
and we add 0.392 as option named recon_loss_weight.
In addition we attached MXNet's tensorboard for plotting.

File: example/capsnet/README.md (hunk 1 above)
Comment (Soonhwan-Kwon) [reply example/capsnet/README.md]: I tested it on MXNet 0.12.1

File: example/capsnet/capsulenet.py (hunk 2 above)
Comment (Soonhwan-Kwon) [reply example/capsnet/capsulenet.py]: Thank you for review and I added 0.392 as option named recon_loss_weight

**Comment by Soonhwan-Kwon** [comment]
@piiswrong Can this example be merged?

**Comment by piiswrong** [comment]
@Soonhwan-Kwon Yes. Could you rebase to master? needs CI to pass

**Comment by Soonhwan-Kwon** [comment]
@piiswrong Thank you for review. I rebased to the master.
